libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/**
+ *
+ * The default tracer implementation. This class implements the basic logic for span lifecycle and its state management.
+ * It also handles tracing context propagation between spans.
+ *
+ *
+ */
+public class DefaultTracer implements Tracer {
+
+    /**
+     * Key for storing current span
+     */
+    public static final String CURRENT_SPAN = "current_span";
+
+    static final String THREAD_NAME = "th_name";
+
+    private final TracingTelemetry tracingTelemetry;
+    private final TracerContextStorage<String, Span> tracerContextStorage;
+
+    /**
+     * Creates DefaultTracer instance
+     *
+     * @param tracingTelemetry tracing telemetry instance
+     * @param tracerContextStorage storage used for storing current span context
+     */
+    public DefaultTracer(TracingTelemetry tracingTelemetry, TracerContextStorage<String, Span> tracerContextStorage) {
+        this.tracingTelemetry = tracingTelemetry;
+        this.tracerContextStorage = tracerContextStorage;
+    }
+
+    @Override
+    public Scope startSpan(String spanName) {
+        Span span = createSpan(spanName, getCurrentSpan());
+        setCurrentSpanInContext(span);
+        addDefaultAttributes(span);
+        return new ScopeImpl(() -> endSpan());
+    }
+
+    @Override
+    public void endSpan() {
```

Comment:
do we really need `endSpan()`  to be public, since we have Scope returned now and consumers should close the scope to end the span? 
Not sure about the intention, but for synchronous flow we don't really need to expose endSpan() when scope closure will take care of it. 
For asynchronous flow, if we are planning to introduce trace preserving action listener, then it should preserve the scope to close it appropriately on response/failure. 

In OpenTelemetry, the lifecycle of Scope if independent of Span. A Span is always associated with a Scope, however, scope can be opened and closed on demand (to stash/restore the context) without closing the Span. Here, the closure of Scope will close the span.

Reply:
> Not sure about the intention, but for synchronous flow we don't really need to expose endSpan() when scope closure will take care of it.

Its an option available for users in whichever way they want to close the span. It would be helpful in cases when on needs to close the span in another method. We needed this while instrumenting search flow

Reply:
I would disagree and commit, so feel free to mark it as resolved or address later if needed. 

Ending span without passing any information about span has downsides like unintentional closure of span/memory leaks in case its not closed, code readability and code maintainability (imagine scenario of refactoring classes where span is closed in a different method using getCurrentSpan() method). This way of closing span is discouraged in OpenTelemetry framework - https://github.com/open-telemetry/opentelemetry-java/blob/ddb47e14c814a1f3e114cbe0c6db50638e9cdf14/context/src/main/java/io/opentelemetry/context/Context.java#L178-L199 and given we are not a generic tracing framework thus I do not see any problem here being  more constrained than OpenTelemetry.
We can go ahead with `@MustBeClosed` annotation for now.

Reply:
I have removed the `endSpan` for now. More details in other comment, closing the thread